UefiPayloadPkg/Library/PciSegmentInfoLibAcpiBoardInfo/PciSegmentInfoLibAcpiBoardInfo.c

```diff
@@ -13,48 +13,98 @@
 
-STATIC PCI_SEGMENT_INFO  mPciSegment0 = {
-  0,  // Segment number
-  0,  // To be fixed later
-  0,  // Start bus number
-  255 // End bus number
-};
+static PCI_SEGMENT_INFO  *PciSegments = 0;
 
 /**
-  Return an array of PCI_SEGMENT_INFO holding the segment information.
+  Find segment info from all root bridges
 
-  Note: The returned array/buffer is owned by callee.
+  @param[in]  PciRootBridgeInfo    Pointer of Universal Payload PCI Root Bridge Info Hob
+  @param[out] NumberOfRootBridges  Number of root bridges detected
 
-  @param  Count  Return the count of segments.
+  @retval     Pointer to the allocated Segment info structure array.
 
-  @retval A callee owned array holding the segment information.
 **/
-PCI_SEGMENT_INFO *
-EFIAPI
-GetPciSegmentInfo (
-  UINTN  *Count
+STATIC PCI_SEGMENT_INFO *
+RetrieveSegmentInfoFromHob (
+  IN  UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES  *PciRootBridgeInfo,
+  OUT UINTN                               *NumberOfRootBridges
   )
 {
-  EFI_HOB_GUID_TYPE  *GuidHob;
-  ACPI_BOARD_INFO    *AcpiBoardInfo;
+  UINTN  Size;
+  UINT8  Index;
 
-  ASSERT (Count != NULL);
-  if (Count == NULL) {
+  ASSERT (PciRootBridgeInfo != NULL);
+  ASSERT (NumberOfRootBridges != NULL);
+  if (PciRootBridgeInfo == NULL) {
     return NULL;
   }
 
-  if (mPciSegment0.BaseAddress == 0) {
-    //
-    // Find the acpi board information guid hob
-    //
-    GuidHob = GetFirstGuidHob (&gUefiAcpiBoardInfoGuid);
-    ASSERT (GuidHob != NULL);
+  *NumberOfRootBridges = PciRootBridgeInfo->Count;
+  if (PciSegments != NULL) {
+    return PciSegments;
+  }
+
+  Size        = PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO);
+  PciSegments = (PCI_SEGMENT_INFO *)AllocatePool (Size);
+  ASSERT (PciSegments != NULL);
+  ZeroMem (PciSegments, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
+
+  //
+  // Create all root bridges with PciRootBridgeInfoHob
+  //
+  for (Index = 0; Index < PciRootBridgeInfo->Count; Index++) {
+    PciSegments[Index].BaseAddress    = PciRootBridgeInfo->RootBridge[Index].Ecam.Base;
+    PciSegments[Index].SegmentNumber  = (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment);
+    PciSegments[Index].StartBusNumber = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Base;
+    PciSegments[Index].EndBusNumber   = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Limit;
+  }
+
+  return PciSegments;
+}
+
+/**
+  Return all the root bridge instances in an array.
+
+  @param Count  Return the count of root bridge instances.
+
+  @return All the root bridge instances in an array.
+          The array should be passed into PciHostBridgeFreeRootBridges()
+          when it's not used.
+**/
+PCI_SEGMENT_INFO *
+EFIAPI
+GetPciSegmentInfo (
+  UINTN  *Count
+  )
+{
+  UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES  *PciRootBridgeInfo;
+  EFI_HOB_GUID_TYPE                   *GuidHob;
+  UNIVERSAL_PAYLOAD_GENERIC_HEADER    *GenericHeader;
 
-    AcpiBoardInfo            = (ACPI_BOARD_INFO *)GET_GUID_HOB_DATA (GuidHob);

```

Comment:
for GetPciSegmentInfo() implementation, could we keep backward compatibility like below?
1. if new UPL_PCI_SEGMENT_INFO_HOB present, use it to report PCI_SEGMENT_INFO.
2. otherwise, use ACPI_BOARD_INFO hob to report PCI_SEGMENT_INFO (old platforms do not use FDT thus would like to use this old HOB)

Reply:
Old platforms will still rely on gUefiAcpiBoardInfoGuid HOB to report PCI_SEGMENT_INFO, they may not create new gUplPciSegmentInfoHobGuid HOB.

We need to support both old/new HOBs to keep backward compatibility, could we enhance the flow like below example code?
```
  if (mPciSegments != NULL) {
    *Count = mCount;
    return mPciSegments;
  }

  if (UplSegmentInfo != NULL) {
    //
    // new model which will rely on RootBridgeInfo and UplSegmentInfo HOBs to update mPciSegments
    //
    PciRootBridgeInfo = Get_RBInfo ();
    UplSegmentInfo    = Get_UPLSegInfo ();
    if ((PciRootBridgeInfo != NULL) && (UplSegmentInfo != NULL)) {
      RetrieveSegmentInfoFromHob(...);
    }
  } else {
    //
    // old model relies on gUefiAcpiBoardInfoGuid and hardcoded values for single segment only. 
    // This is only for backward compatibility, new platforms should adopt new model even in single segment cases.
    //
    mPciSegments = (PCI_SEGMENT_INFO *)AllocatePool (sizeof (PCI_SEGMENT_INFO));
    ASSERT (mPciSegments != NULL);
    if (mPciSegments != NULL) {
      GuidHob = GetFirstGuidHob (&gUefiAcpiBoardInfoGuid);
      ASSERT (GuidHob != NULL);
      if (GuidHob != NULL) {
        AcpiBoardInfo = (ACPI_BOARD_INFO *)GET_GUID_HOB_DATA (GuidHob);
        mPciSegment->SegmentNumber = 0;
        mPciSegment->BaseAddress = AcpiBoardInfo->PcieBaseAddress;
        mPciSegment->StartBusNumber = 0;
        mPciSegment->EndBusNumber =  0;
        mCount = 1;

      }
    }    
  }
  *Count = mCount;
  return mPciSegments; // either 0 or valid PCI_SEGMENT_INFO data structure pointer
```

Reply:
done please verify on your platform.